Autopilot/PathManager/Src/waypointManager.cpp

```diff
@@ -0,0 +1,357 @@
+/**
+ * Waypoint Manager Methods and Helpers Implementation
+ * Author: Dhruv Rawat
+ * Created: November 2020
+ * Last Updated: November 2020 (Dhruv)
+ */
+
+#include "waypointManager.hpp"
+
+
+/*** INITIALIZATION ***/
+
+
+WaypointManager::WaypointManager(_PathData *initialWaypoints, int numberOfWaypoints, _PathData *currentLocation) {
+    
+    homeBase = _PathData(); //Initializes homebase first so there are no undefined behaviour issues
+    homeBase = *currentLocation;
+
+    numWaypoints = numberOfWaypoints;
+    nextFilledIndex = 0;
+    nextAssignedId = 0;
+    
+    // Initializes the waypointBuffer array
+    for (int i = 0; i < numWaypoints; i++) {
+        // Links waypoints together
+        if (i == 0) { // If first waypoint, link to next one only
+            initialWaypoints[i].next = &initialWaypoints[i+1]; // Passes address so any updates in initialWaypoints[i+1] is reflected in initialWaypoints[i].next
```

Comment:
in doubly linked list, general protocol is first_node.previous = NULL; and last_node.next = NULL